ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java

```diff
@@ -483,12 +471,12 @@ protected Boolean findNextCompactionAndExecute(boolean computeStats) throws Inte
        * multiple statements in it (for query based compactor) which is not supported (and since
        * this case some of the statements are DDL, even in the future will not be allowed in a
        * multi-stmt txn. {@link Driver#setCompactionWriteIds(ValidWriteIdList, long)} */
-      compactorTxnId = msc.openTxn(ci.runAs, TxnType.COMPACTION);
+      compactionTxn.open(ci);
 
-      heartbeater = new CompactionHeartbeater(compactorTxnId, fullTableName, conf);
+      heartbeater = new CompactionHeartbeater(compactionTxn.getTxnId(), fullTableName, conf);
```

Comment:
Maybe change compactionHeartbeater to use the new class instead of Id?

Reply:
Done